### PR TITLE
ATB-1425 : Added to metric which calculates time difference in seconds from when an account was suspended to when it got unsuspended

### PR DESCRIPTION
**Pull request by TaherKapasi**
### What changed
Added to metric which calculates time difference in seconds from when an account was suspended to when it got unsuspended

### Why did it change
To measure the time taken (seconds) for a suspended account to become unsuspended, either through user action or explicitly with an unsuspend intervention.

### Issue tracking

- [ATB-1425](https://govukverify.atlassian.net/browse/ATB-1425)

## Testing
Generated and viewed metrics for following use cases.

Suspend and unsuspend 
Suspend with Password Reset then Password Reset Successful
Suspend with ID Reset then ID Reset Successful
Suspend with Password Reset and ID Reset then Password Reset Successful then ID Reset Successful

Screenshot 2024-01-19 at 22 17 07
Screenshot 2024-01-19 at 22 15 31
Screenshot 2024-01-19 at 22 08 46

## Checklists
- [x] Did not commit any not-required changes to the `src/infra/**/samconfig.toml`
- [x] Tested changes and included test evidence in the PR, if appropriate
- [ ] Included all required tags and other properties for any new resources in the SAM template
- [ ] Ensured that any new resources in the SAM Template follow appropriate naming conventions
- [ ] Ensured that naming of new resources is compatible with deploying multiple stacks with custom stack names during development
- [ ] Ensured that no log lines include PII or other sensitive data
- [x] Implemented unit testing for any new logic implemented, if appropriate
- [x] Ensured that all commits in this PR are signed
- [x] Ensured appropriate code coverage is maintained by unit tests
- [x] Checked SonarCube and ensured no code smells were added

[ATB-1425]: https://govukverify.atlassian.net/browse/ATB-1425?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by liamquayle — APPROVED** [review]

File: src/commons/metrics-helper.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+export function publishTimeToResolveMetrics(
+  oldState: StateDetails,
+  currentState: StateDetails,
+  oldAppliedAt: number,
+  currentAppliedAt: number,
+  eventName: string,
+) {
+  const timeDifferenceAsSeconds = Math.floor((currentAppliedAt - oldAppliedAt) / 1000);
+
+  if (oldState.suspended && !currentState.suspended) {
```
Comment (chiarapascucci-gds) [inline src/commons/metrics-helper.ts:45]: hey Taher - where an user action event is causing the user to go from (for example) Suspended Psw Reset to Unsuspended, then we would be logging this metric twice, I think? Once with type `suspension` and once with type `resetPassword` - is this the intended behaviour?

File: src/handlers/interventions-processor-handler.ts [hunk 2]
```diff
@@ -104,8 +104,16 @@ async function processSQSRecord(record: SQSRecord) {
     recordBody,
     statusResult.interventionName,
   );
-  logger.debug('processed requested event, sending update request to dynamo db');
+  logger.debug('Updating user status', { userId, partialCommandInput });
   await service.updateUserStatus(userId, partialCommandInput);
+  publishTimeToResolveMetrics(
+    currentAccountState,
+    statusResult.stateResult,
+    itemFromDB?.appliedAt ? itemFromDB.appliedAt : currentTimestamp.milliseconds,
```
Comment (chiarapascucci-gds) [inline src/handlers/interventions-processor-handler.ts:112]: linting is complaining about the ternary here.. I think `itemFromDB?.appliedAt || currentTimestamp.milliseconds` is equivalent and should satisfy the liniting

File: src/commons/metrics-helper.ts (hunk 1 above)
Comment (TaherKapasi) [reply src/commons/metrics-helper.ts:45]: Hi @chiarapascucci-gds - yes - one metric to capture the time to resolve with a user led action and another to capture the time to unsuspend an account (either through user led action or when fraud remove suspension explicitly) - we may get only one metric published if the account is still suspended

File: src/commons/metrics-helper.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  if (
+    (eventName === EventsEnum.AUTH_PASSWORD_RESET_SUCCESSFUL ||
+      eventName === EventsEnum.AUTH_PASSWORD_RESET_SUCCESSFUL_FOR_TEST_CLIENT) &&
+    oldState.resetPassword &&
+    !currentState.resetPassword
+  ) {
+    logAndPublishMetric(MetricNames.TIME_TO_RESOLVE, noMetadata, timeDifferenceAsSeconds, { type: 'resetPassword' });
+  }
+
+  if (eventName === EventsEnum.IPV_IDENTITY_ISSUED && oldState.reproveIdentity && !currentState.reproveIdentity) {
```
Comment (chiarapascucci-gds) [inline src/commons/metrics-helper.ts:58]: more of a general question/comment. Where the user received a psw&id reset required intervention (06), when we log the time to resolve of whichever side gets resolved second (I am assuming ipv identity issued most likley?), then this Time_to_resolve value will include the time to resolve of the first one as well, which may skew what we are trying to measure. Not an issue probably, but something to be aware of, maybe?

**Review by chiarapascucci-gds — APPROVED** [review]

File: src/commons/metrics-helper.ts (hunk 3 above)
Comment (TaherKapasi) [reply src/commons/metrics-helper.ts:58]: Hi @chiarapascucci-gds  - yes this is expected:  So FA fires an 06 at 1st jan 00:00.  User resolves password at 01:00,  TimeToResolve (resetPassword) would have a metric of 3600.  Then the user resets their identity by 02:00 emitting a TimeToResolve (reproveIdentity) 7200 and  TimeToResolve (suspension) 7200.

**Pull request merged**